prow/cmd/pipeline/controller.go

```diff
@@ -0,0 +1,684 @@
+			npj.Status.StartTime = c.now()
+		}
+		if npj.Status.CompletionTime.IsZero() && finalState(state) {
+			now := c.now()
+			npj.Status.CompletionTime = &now
+		}
+		npj.Status.State = state
+		npj.Status.Description = msg
+		logrus.Infof("Update %s/%s: %s -> %s", prowJob, npj.Name, haveState, state)
+		if _, err := c.updateProwJob(npj); err != nil {
+			return fmt.Errorf("update prow status: %v", err)
+		}
+	}
+	return nil
+}
+
+// isReported check if a prow job was already reported
+func isReported(status *prowjobv1.ProwJobStatus) bool {
+	if status != nil && status.PrevReportStates != nil && (status.PrevReportStates["github-reporter"] == prowjobv1.PendingState) {
+		return true
+	}
+	return false
+}
+
+// finalState returns true if the prowjob has already finished
+func finalState(status prowjobv1.ProwJobState) bool {
+	switch status {
+	case "", prowjobv1.PendingState, prowjobv1.TriggeredState:
+		return false
+	}
+	return true
+}
+
+// description computes the ProwJobStatus description for this condition or falling back to a default if none is provided.
+func description(cond duckv1alpha1.Condition, fallback string) string {
+	switch {
+	case cond.Message != "":
+		return cond.Message
+	case cond.Reason != "":
+		return cond.Reason
+	}
+	return fallback
+}
+
+const (
+	descScheduling = "scheduling"
+	descRunning    = "running"
+	descSucceeded  = "succeeded"
+	descFailed     = "failed"
+	descUnknown    = "unknown status"
+)
+
+// prowJobStatus returns the desired state and description based on the pipeline status
+func prowJobStatus(ps pipelinev1alpha1.PipelineRunStatus) (prowjobv1.ProwJobState, string) {
+	pcond := ps.GetCondition(duckv1alpha1.ConditionSucceeded)
+	if pcond == nil {
+		return prowjobv1.TriggeredState, descScheduling
+	}
+	cond := *pcond
+	switch {
+	case cond.Status == untypedcorev1.ConditionTrue:
+		return prowjobv1.SuccessState, description(cond, descSucceeded)
+	case cond.Status == untypedcorev1.ConditionFalse:
+		return prowjobv1.FailureState, description(cond, descFailed)
+	case cond.Status == untypedcorev1.ConditionUnknown:
+		return prowjobv1.PendingState, description(cond, descRunning)
+	}
+
+	logrus.Warnf("Unknown condition %#v", cond)
+	return prowjobv1.ErrorState, description(cond, descUnknown) // shouldn't happen
+}
+
+// ownerRef builds an owner reference from prow job
+func ownerRef(pj prowjobv1.ProwJob) metav1.OwnerReference {
+	return metav1.OwnerReference{
+		APIVersion: prowjobv1.SchemeGroupVersion.String(),
+		Kind:       prowJob,
+		Name:       pj.Name,
+		UID:        pj.UID,
+	}
+}
+
+// pipelineMeta builds the pipeline metadata from prow job definition
+func pipelineMeta(pj prowjobv1.ProwJob) metav1.ObjectMeta {
+	labels, annotations := decorate.LabelsAndAnnotationsForJob(pj)
+	return metav1.ObjectMeta{
+		Annotations:     annotations,
+		Name:            pj.Name,
+		Namespace:       pj.Spec.Namespace,
+		Labels:          labels,
+		OwnerReferences: []metav1.OwnerReference{ownerRef(pj)},
```

Comment:
Can we please drop the owner references? This does not work with cross cluster resources. AKA if you create a resource in a foreign cluster with an owner reference to a local resource, the foreign cluster will immediately delete the object because resource name does not exist in the foreign cluster.

I'm pretty sure later commits of the build controller dropped owner references, this is why

Reply:
I removed it. This was my addition, I guess. I didn't know about this behaviour cross clusters. Thanks for explaining the details.